react/lesson1.md

```diff
@@ -36,175 +91,431 @@ cd my-app/
 
-  handleChange(event) {
-    this.setState({website: event.target.value});
+```javascript
+  selectRandomButton = () => {
+    const randomIndex = this.getRandomInt(0, this.state.difficulty);
+    this.setState({
+      selectedButton: randomIndex
+    })
   }
-  render() {
-    return (
-      <div className="App">
-        <Organisation
-            name="Refugee Support Network"
-            website={this.state.website}
-        />
-        <input type="text" value={this.state.website} onChange={this.handleChange.bind(this)} />
-      </div>
-    );
+
+  getRandomInt(min, max) {
+    min = Math.ceil(min);
+    max = Math.floor(max);
+    return Math.floor(Math.random() * (max - min)) + min;
   }
-}
+```
 
-export default App;
+Modify the `buildGame()` function so it only selects the button that's been marked on the state:
 
+```javascript
+buttons.push(<GameButton selected={this.state.selectedButton === i}/>)
 ```
 
-- [State Exercise](https://codepen.io/kabaros/pen/NdRNmm?editors=0010#0)
+# Add a timer
+
+Let's a delay between the time we press "Start game" and one of the buttons getting selected:
+
+```javascript
+  startGame = () => {
+    setTimeout(this.selectRandomButton, 2000);
+  }
+
+  ///render()
+  <button onClick={this.startGame}>Start game</button> 
+```
+
+# Add a click handler to our buttons
+
+```javascript
+<GameButton onClick={this.gameButtonClick} selected={this.state.selectedButton === i}/>
+```
+
+And our click handler:
+
+```javascript
+gameButtonClick = (event) => {
+  const className = event.target.className;
+  if(className.includes('selected')) {
+    alert("GOOD JOB");
+  }
+};
+```
+
+Click on "Start game" and wait for one of the buttons to become selected. Once a button becomes selected, click on it. Hmm, it still doesn't work.
+
+That's because inside our `GameButton` component, we are doing nothing to handle the `onClick` prop we're sending in. Let's change that!
+
+Inside `gameButton.js`:
+
+```javascript
+<button onClick={props.onClick}...
+```
+
+# Calculate how fast you click on a button
+
+This is the point where we add a challenge and turn our app into a game. We need to calculate how fast the user reacts to a button being selected. 
+
+Add a start and end time to the state:
+
+```javascript
+  constructor() {
+    super();
+    this.state = {
+      difficulty: 10,
+      selectedButton: -1,
+      startTime: 0
+    };
+  }
+
+  // When a random gets selected set the START TIME
+  selectRandomButton = () => {
+    const randomIndex = this.getRandomInt(0, this.state.difficulty);
+    this.setState({
+      selectedButton: randomIndex,
+      startTime: new Date().getTime()
+    })
+  }
+
+  // When the user manages to click on the selected button, display the time passed start and end
+  gameButtonClick = (event) => {
+    const className = event.target.className;
+    if(className.includes('selected')) {
```

Comment:
This feels like an antipattern - We shouldn't be looking at the state of the DOM to determine whether a clicked element has been selected.

I haven't pulled the repo down locally and ran the app so I can't offer a nice alternative right now but it feels like a closure/bound function that returns this above function, and has the number of the clicked button would be nicer.

Reply:
You are totally right! Although I might leave as it is, going into how the closure works in this case would be a bit too much I think?

Reply:
Yeh - this could be better of as a bit of a refactor to avoid both a closure and this - i'll pull the repo tomorrow and take a look.